assume/reinforcement_learning/algorithms/matd3.py

```diff
@@ -24,183 +25,111 @@ def __init__(
         target_policy_noise=0.2,
         target_noise_clip=0.5,
     ):
-        super().__init__()
-
-        self.env = env
+        self.learning_role = learning_role
         self.learning_rate = learning_rate
         self.learning_starts = learning_starts
         self.batch_size = batch_size
         self.gamma = gamma
         self.tau = tau
 
-        self.train_freq = len(env.snapshots) if train_freq == -1 else train_freq
+        self.train_freq = train_freq
         self.gradient_steps = (
             self.train_freq if gradient_steps == -1 else gradient_steps
         )
+
         self.policy_delay = policy_delay
         self.target_noise_clip = target_noise_clip
         self.target_policy_noise = target_policy_noise
 
-        self.rl_agents = [
-            agent for agent in env.rl_powerplants + env.rl_storages if agent.learning
-        ]
-        self.n_rl_agents = len(self.rl_agents)
+        self.obs_dim = self.learning_role.obs_dim
+        self.act_dim = self.learning_role.act_dim
 
-        self.obs_dim = env.obs_dim
-        self.act_dim = env.act_dim
+        self.device = self.learning_role.device
+        self.float_type = self.learning_role.float_type
 
-        self.device = env.device
-        self.float_type = env.float_type
+        self.unique_obs_len = 8
 
-        self.buffer = ReplayBuffer(
-            buffer_size=int(buffer_size),
-            obs_dim=self.obs_dim,
-            act_dim=self.act_dim,
-            n_rl_units=self.n_rl_agents,
-            device=self.device,
-        )
+    def update_policy(self):
+        logger.info(f"Updating Policy")
+        n_rl_agents = len(self.learning_role.rl_units)
+        for _ in range(self.gradient_steps):
+            i = 0
+            strategy: LearningStrategy
+            for u_id, strategy in self.learning_role.rl_units.items():
+                critic_target = self.learning_role.target_critics[u_id]
+                critic = self.learning_role.critics[u_id]
+                actor = self.learning_role.rl_units[u_id].actor
```

Comment:
Okay, so you solved this now by giving all units to the learning role and the learning role is in the algorithm and hence you can access the actor directly. Seems like a good work around for now. 

Why do we still have the if i%100==0: though I thought this is handeld by the sheduling of the recuurent task?

Reply:
this ensures that when you have many agents, they don't learn on the same sample from the buffer. we can remove this for now, but it improved the performance for me

Reply:
also, the main point was not to compute next actions so many times as this is the most intensive task and requires a ton of time

Reply:
Ah okay, I misinterpreted this then